### DISCUSSION TITLE
How to remove no-op exported functions?

### DISCUSSION THREAD

**Discussion by orsinium** [WebAssembly/binaryen#6501] [Q&A]
Quite often, after all optimizations, wasm-opt leaves exported functions empty:

```
(module
  (type (func))
  (func (type 0)
    nop)
  (export "example" (func 0)))
```

Is there a way to tell wasm-opt that it is also safe to remove the exported functions if they are empty? I know for sure that the target runtime handles missed exports correctly.

**Comment by kripken** [comment]
wasm-metadce can remove exports that you specify:

https://github.com/WebAssembly/binaryen/wiki/Pruning-unneeded-code-in-wasm-files-with-wasm-metadce#example-pruning-exports

But it won't automatically remove empty ones, so you'd need to find out which are empty and give it those as a list. That could be fairly easily done by parsing the wat file though, I think.

**Reply by orsinium** [reply]
> That could be fairly easily done by parsing the wat file though

I build from Go code, which then gets optimized by LLVm and binaryen. So, I'd need to parse the resulting WASM file and add one more post-processing step, which will run custom code with parsing wasm, analyzing it, removing some functions, and then dumping it back. I hoped for a simpler out-of-the-box solution.

Should I open a feature-request?

**Reply by kripken** [reply]
I think a feature is worth considering. Specifically we could add a wasm-metadce flag to consider empty functions unused even if exported. But I am not sure how simple that is to implement and whether that is generally useful, which would be two things we'd take into account when considering a new feature.

**Reply by orsinium** [reply]
https://github.com/WebAssembly/binaryen/issues/6506
